### PR TITLE
gluon conv rnns

### PR DESCRIPTION
**Pull request by szha**
@sxjscience

### CODE REVIEW

File: python/mxnet/gluon/rnn/rnn_cell.py [hunk 1]
```diff
@@ -777,3 +785,329 @@ def unroll(self, length, inputs, begin_state=None, layout='NTC', merge_outputs=N
...
+        Dilation of Convolution operator in input-to-state transitions.
+    activation : str or Block
+        default LeakyReLU(alpha=0.2)
+        Type of activation function.
+    prefix : str, default 'ConvLSTM_'
+        Prefix for name of layers (and name of weight if params is None).
+    params : RNNParams, default None
+        Container for weight sharing between cells. Created if None.
+    forget_bias : bias added to forget gate, default 1.0.
+        Jozefowicz et al. 2015 recommends setting this to 1.0
```
Comment (sxjscience) [inline python/mxnet/gluon/rnn/rnn_cell.py]: I find that the forget_bias is not used. Also, for ConvLSTM, there seems to be not much difference for different forget_bias. I think it's safe to simply remove this comment.

File: tests/python/unittest/test_gluon_rnn.py [hunk 2]
```diff
@@ -192,6 +196,59 @@ def test_rnn_cells():
...
+    net.add(gluon.rnn.ConvGRUCell(100, input_shape=(100, 5, 7)))
+    check_rnn_forward(net, mx.nd.ones((8, 3, 200, 5, 7)))
+
+
+def test_convrnn():
+    cell = gluon.rnn.ConvRNNCell(hidden_size=10, input_shape=(3, 16, 10),
+                                 h2h_kernel=(3, 3), h2h_dilate=(1, 1),
+                                 i2h_kernel=(3, 3), i2h_stride=(1, 1),
+                                 i2h_pad=(1, 1), i2h_dilate=(1, 1),
+                                 prefix='rnn_')
```
Comment (sxjscience) [inline tests/python/unittest/test_gluon_rnn.py]: I find that the `conv_layout` has not been tested. It would be safer to test it.

Comment (szha) [inline tests/python/unittest/test_gluon_rnn.py]: Good call. I was able to discover some issues when testing it.

**Comment by szha** [comment]
@dsqx71 it seems that in the symbolic RNN API, the conv_layout is not passed to the convolution layers, and some of the variable layouts are hard-coded according to 'NCHW' layout in #6832. Would you refer to the fixes in this PR and fix the one in symbolic RNN API?

**Comment by dsqx71** [comment]
@szha  Thanks. I will fix it.

File: python/mxnet/gluon/rnn/rnn_cell.py [hunk 3]
```diff
@@ -784,3 +792,378 @@ def unroll(self, length, inputs, begin_state=None, layout='NTC', merge_outputs=N
...
+        Initializer for the bias vector.
+    activation : str or Block
+        default LeakyReLU(alpha=0.2)
+        Type of activation function.
+    prefix : str, default 'conv_rnn_'
+        Prefix for name of layers (and name of weight if params is None).
+    params : RNNParams, default None
+        Container for weight sharing between cells. Created if None.
+    """
+    def __init__(self, hidden_size, conv_layout='NCHW', input_shape=(3, 224, 224),
```
Comment (piiswrong) [inline python/mxnet/gluon/rnn/rnn_cell.py]: input_shape shouldn't be consted like this. layout should be put at the end. put activation in the front

Comment (szha) [inline python/mxnet/gluon/rnn/rnn_cell.py]: How do I defer the shape_info initialization? It depends on input.

**Comment by szha** [comment]
Somehow the tests fail in MKLConcatOp in the MKL variant. @ykim362 any idea what this is?

**Comment by szha** [comment]
@ykim362 looks like a problem in MKL concat. The following code causes the same error when using MKL version to do 5 dim concat
```
mx.nd.concat(*[mx.nd.ones((1,1,1,1,1)), mx.nd.ones((1,1,1,1,1))], dim=1)
```

**Comment by ykim362** [comment]
@szha I am sorry for late reply. I am out of office this week. MKL's concat does not support general dimensions. I might need to change the MXNet-MKL integration code so that the CPU code call the default library when MKL doesn't support the specific dimensions. I will do that when I return to the office next week.

**Comment by szha** [comment]
@sxjscience @piiswrong this PR is ready for review and merge.

**Comment by piiswrong** [comment]
Let's make input_shape = () by default and raise error when begin_state is called without specifying input_shape.

Users can still create begin state by them selves and ignore input_shape

**Comment by adstraw** [comment]
MKL supports 2D/3D/4D concatenation only.  5D concatenation must not go through MKL.  See this pull request which makes that change:

https://github.com/apache/incubator-mxnet/pull/7868

File: python/mxnet/gluon/contrib/rnn/conv_rnn_cell.py [hunk 4]
```diff
@@ -0,0 +1,971 @@
...
+
+    def state_info(self, batch_size=0):
+        raise NotImplementedError("_BaseConvRNNCell is abstract class for convolutional RNN")
+
+    def hybrid_forward(self, F, inputs, states):
+        raise NotImplementedError("_BaseConvRNNCell is abstract class for convolutional RNN")
+
+
+class _ConvRNNCell(_BaseConvRNNCell):
+    def __init__(self, hidden_channels, input_shape, activation,
```
Comment (piiswrong) [inline python/mxnet/gluon/contrib/rnn/conv_rnn_cell.py]: input_shape first.

File: python/mxnet/gluon/contrib/rnn/conv_rnn_cell.py [hunk 5]
```diff
@@ -0,0 +1,971 @@
...
+        self._activation = activation
+
+        # Convolution setting
+        assert all(isinstance(spec, int) or len(spec) == dims
+                   for spec in [i2h_kernel, i2h_stride, i2h_pad, i2h_dilate,
+                                h2h_kernel, h2h_dilate]), \
+               "For {dims}D convolution, the convolution settings can only be either int " \
+               "or tuple of length {dims}".format(dims=dims)
+
+        self._i2h_kernel = i2h_kernel if isinstance(i2h_kernel, tuple) else (i2h_kernel,) * dims
```
Comment (piiswrong) [inline python/mxnet/gluon/contrib/rnn/conv_rnn_cell.py]: see checks in conv layers. Use can also use list

File: python/mxnet/gluon/contrib/rnn/conv_rnn_cell.py [hunk 6]
```diff
@@ -0,0 +1,971 @@
...
+        Initializer for the recurrent convolution bias vectors.
+    conv_layout : str, default 'NCW'
+        Layout for all convolution inputs, outputs and weights. Options are 'NCW' and 'NWC'.
+    prefix : str, default 'conv_rnn_'
+        Prefix for name of layers (and name of weight if params is None).
+    params : RNNParams, default None
+        Container for weight sharing between cells. Created if None.
+    """
+    def __init__(self, hidden_channels, input_shape, activation='tanh',
+                 i2h_kernel=(3,), i2h_stride=(1,), i2h_pad=(1,), i2h_dilate=(1,),
```
Comment (piiswrong) [inline python/mxnet/gluon/contrib/rnn/conv_rnn_cell.py]: Why is there a default value for kernel?

Comment (szha) [reply python/mxnet/gluon/contrib/rnn/conv_rnn_cell.py]: state what you'd like please

Comment (szha) [reply python/mxnet/gluon/contrib/rnn/conv_rnn_cell.py]: https://github.com/apache/incubator-mxnet/blob/master/python/mxnet/rnn/rnn_cell.py#L1185-L1196

**Comment by szha** [comment]
@mbaijal CI doesn't seem to be running.

File: python/mxnet/gluon/contrib/rnn/conv_rnn_cell.py (hunk 6 above)
Comment (piiswrong) [reply python/mxnet/gluon/contrib/rnn/conv_rnn_cell.py]: I don't know. But considering Conv2D doesn't set default value for kernel, what's the reason for adding one here? Also default padding is 0 for Conv2D.

Comment (szha) [reply python/mxnet/gluon/contrib/rnn/conv_rnn_cell.py]: @sxjscience what would you recommend?

**Comment by mbaijal** [comment]
@szha, Do you mean the current builds that are failing in builds.apache or the above checks in github are not working?

File: python/mxnet/gluon/contrib/rnn/conv_rnn_cell.py (hunk 6 above)
Comment (piiswrong) [reply python/mxnet/gluon/contrib/rnn/conv_rnn_cell.py]: Either way this needs to be better documented.

**Comment by szha** [comment]
@mbaijal the status is not shown in the PR.

**Comment by mbaijal** [comment]
@szha, I have also noticed some irregularities here. I will look into this. Thanks for pointing it out.

File: python/mxnet/gluon/contrib/rnn/conv_rnn_cell.py (hunk 6 above)
Comment (sxjscience) [reply python/mxnet/gluon/contrib/rnn/conv_rnn_cell.py]: I agree that we should remove the default.

Comment (sxjscience) [reply python/mxnet/gluon/contrib/rnn/conv_rnn_cell.py]: I think we should remove the default value of i2h_kernel and h2h_kernel and keep the default value of i2h_pad, i2h_stride,... h2h_dilate to be the same as that of the convolution layer.

Comment (szha) [reply python/mxnet/gluon/contrib/rnn/conv_rnn_cell.py]: Will do

Comment (szha) [reply python/mxnet/gluon/contrib/rnn/conv_rnn_cell.py]: I have updated the parameters. In addition I made two other changes:
1. i2h and h2h strides are tied. Otherwise i2h and h2h may not match in size
2. activation is moved to last because it has default value of tanh, which can't precede the required parameters.

**Comment by szha** [comment]
Per offline discussion with @sxjscience, I removed strides support to avoid complication. This is consistent with what tf does
